### PR TITLE
Delay all imports in `napari.__init__` behind module level `napari.__getattr__`

### PR DESCRIPTION
**Pull request by tlambert03**
# Description
following on #2371 and [this](https://napari.zulipchat.com/#narrow/stream/212875-general/topic/slow.20pythonw) conversation in zulip,  this PR provides an example of what our `__init__` could look like using module `__getattr__`.  `import napari` is nearly instantaneous, which is useful when being called as `python -m napari` or `napari` from CLI... provided we hide the ensuing imports from napari within `__main__` as shown here.

Since #2535 fixed the greedy import of plugins, the need is a little less dire, but here it is if anyone wants to try it out.  cc @jni

### CODE REVIEW

**Comment by goanpeca** [comment]
I would favor this approach and give it a bump @tlambert03 .

Specially since it would be very useful in getting https://github.com/napari/napari/pull/2760 in, and working as expected!

Any thoughts @napari/core-devs ?

**Review by goanpeca — APPROVED** [review]
I would favor this approach and give it a bump @tlambert03 .

I tested it and works as expected :) 

Now, specially since it would be very useful in getting #2760 in, and working as expected!

Any thoughts @napari/core-devs ?

**Review by sofroniewn — APPROVED** [review]
I'm good with this change! Will move to the awaiting merge pipeline to see if @jni wants to comment before final merge

**Comment by sofroniewn** [comment]
Oh actually, looks like we've got lots of test failings, more than normal! Probably need to sort that out before merge

**Comment by tlambert03** [comment]
fixed (locally at least)

File: napari/utils/_magicgui.py [hunk 1]
```diff
@@ -345,6 +351,7 @@ def add_layer_to_viewer(
     ... def make_layer() -> napari.layers.Image:
     ...     return napari.layers.Image(np.random.rand(64, 64))
     """
+    print("add_layer_to_viewer", result)
```
Comment (jni) [inline napari/utils/_magicgui.py]: These prints need to go... ;)

**Review by jni — APPROVED** [review]
A couple of stray prints still remain, otherwise this is great, thank you @tlambert03!

**Comment by goanpeca** [comment]
I think we still need to fix some tests :)

**Comment by tlambert03** [comment]
removed prints, and also I added an `__init__.pyi` file (otherwise, an IDE won't know what `napari.Viewer` is, for instance)

All of the test fails that I could see in the previous run were notification-manager/threading issues.

**Comment by sofroniewn** [comment]
All tests now, passing this looks great! Will merge!!

**Pull request merged**